README.md

```diff
@@ -0,0 +1,171 @@
+ ```
+type Process struct {
+	clientID string    // client ID retrieved from the client certificate 
+	cmd      *exec.Cmd // process object in Go
+	logfile  string    // path to the file containing combined process standard and error outputs
+	status   string    // process status
+}
+ ```
+
+The library exposes singleton process manager object holding all aforementioned structures:
+```
+const (
+	LogDir = "/tmp/gravitest/proc/logs"
+
+	PermStart  = 0x01
+	PermStop   = 0x02
+	PermStatus = 0x04
+	PermStream = 0x08
+
+	StatusInit    = "initializing"
+	StatusRunning = "running"
+	StatusFailed  = "failed"
+	StatusStopped = "stopped"
+)
+
+type ProcManager struct {
+	// process table [process UID : Process]
+	procs     map[uint64]*Process
+	procMutex sync.Mutex
+	// permission table [client ID : permission bitmap]
+	perm map[string]int
+
+	// process UID
+	uid uint64
+}
+```
+
+### API implementation
+
+The API proto spec is declared in [proto/worker.proto](./proto/worker.proto)
+
+`StartProcess`:
+ - Input: executable name and optional list of arguments.
+ - Output: process UID.
+ - Action:
+   1. verify client authorization to call this API.
+   2. generate a new process UID and create a `Process` object.
+   3. create process logfile `<log directory>/proc-<UID>.log` and redirect combined standard and error output of the process to the logfile.
+   4. add a new entry in the process table.
+   5. start the process by calling `process.cmd.Start()`, and set status to `running`.
+   6. set status to `stopped` once the process is terminated.
+   *note:* steps 5 and 6 are executed asynchronously in a go-routine.
+
+`StopProcess`:
+ - Input: process UID.
+ - Output: none.
+ - Action:
+   1. verify client authorization to call this API.
+   2. if the process is not found in the process table, return `process not found` error.
+   3. if the process is running, terminate it by calling `process.cmd.Process.Kill()`
+
+`GetProcessStatus`:
+ - Input: process UID.
+ - Output: process status.
+ - Action: if the process is in the process table, return process status from the `Process` object. Otherwise return `process not found` error.
+
+`stream-output`:
+ - Input: process UID.
+ - Output: stream of the process output file.
+ - Action:
+   1. verify client authorization to call this API.
+   2. get logfile path from the `Process` object.
+   3. open logfile and start streaming it (gRPC server-side streaming)
+
+### Client
+
+The client is a console application performing the following steps:
+ - initiates gRPC connection over mTLS with the server. The TLS certificate could be passed in the command line or specified via environment variable.
+ - provides CLI to invoke API calls with the server.
+ - produces necessary output.
+
+CLI usage examples:
+```
+$ ./client start -c "ping 8.8.8.8"
+Process UID: 1
+
+./client status -i 1
```

Comment:
If the `-i ` is always required meybe it works to have it as command argument instead of command flag ?

For example you can invoke ping call by 
```
ping 8.8.8.8
```
instead of 
```
ping -d 8.8.8.8
```

where calling `ping` without destination address don't make any sense. 

Also how the CLI will allow to provide a client certificate ?

Reply:
agree, let me change that.